File: List/矩形覆盖.py
# -*- coding:utf-8 -*-
class Solution:
    def __init__(self):
        self.cache = dict()
        self.cache[0] = 0
        self.cache[1] = 1
        self.cache[2] = 2
    def rectCover(self, number):
        # write code here
        if number <= 2:
            return self.cache[number]
        if (number-1) in self.cache and (number-2) in self.cache:
            return self.cache[number-1] + self.cache[number-2]
        self.cache[number-1] = self.rectCover(number-1)
        self.cache[number-2] = self.rectCover(number-2)
        return self.cache[number-1] +  self.cache[number-2]

File: List/矩形覆盖.cpp
class Solution {
public:
    int rectCover(int number) {
        int results[] = {0,1,2};
        if(number <= 2){
            return results[number];
        }
        int minusOne = 2;
        int minusTwo = 1;
        int res = 0;
        for(int i=3; i<=number; i++){
            res = minusOne + minusTwo;
            minusTwo = minusOne;
            minusOne = res;
        }
        return res;
    }
};